src/common/options.cc

```diff
@@ -8010,6 +8010,24 @@ std::vector<Option> get_mds_options() {
     .set_flag(Option::FLAG_RUNTIME)
     .set_long_description("This is the order of magnitude difference (in base 2) of the internal liveness decay counter and the number of capabilities the session holds. When this difference occurs, the MDS treats the session as quiescent and begins recalling capabilities."),
 
+    Option("mds_session_cap_acquisition_decay_rate", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(10)
+    .set_description("decay rate for session readdir caps leading to readdir throttle")
+    .set_flag(Option::FLAG_RUNTIME)
+    .set_long_description("The half-life for the session cap acquisition counter of caps acquired by readdir. This is used for throttling readdir requests from clients slow to release caps."),
+
+    Option("mds_session_cap_acquisition_throttle", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(500000)
```

Comment:
@batrick Should this be relative to mds_max_caps_per_client instead of absolute number ?

Reply:
I was thinking about that but couldn't decide. An argument could be made that it be:

`(mds_session_max_caps_throttle_ratio-1) * mds_max_caps_per_client)`

I think that'd be reasonable actually.

Reply:
I too am not able to decide. I am thinking loud here. 

I think that looses the ability/flexibility to tune the throttling correctly. It ties the throttling point of decay counter to number of caps acquired by the client. 

We need to answer following questions.

1. What percentage of caps acquired by readdir should trigger readdir throttling ?        
2. What percentage of caps acquired in general (others + readdir) above max_caps_per_client should trigger readdir throttling ?
3. Does both percentages should be same ?

With above approach, both will be 10% of max_caps_per_client. So with current defaults 10k caps by readdir and 1.1M total caps acquired by the client will trigger throttling. I think with readdir on large directory kind of workload, decay counter condition will mostly be true and number of total caps ends up deciding readdir throttling. 

In a hypothetical scenario where readdir share of caps is 10k and caps acquired other than readdir 1M, we end up throttling readdir instead of throttling others? Is this ok ?

Reply:
> I too am not able to decide. I am thinking loud here.
> 
> I think that looses the ability/flexibility to tune the throttling correctly. It ties the throttling point of decay counter to number of caps acquired by the client.
> 
> We need to answer following questions.
> 
>     1. What percentage of caps acquired by readdir should trigger readdir throttling ?
> 
>     2. What percentage of caps acquired in general (others + readdir) above max_caps_per_client should trigger readdir throttling ?
> 
>     3. Does both percentages should be same ?

I don't think there's a clear answer for this without testing. I think keeping the separate config gives us the flexibility to fix this in a production cluster so it makes sense to keep it as-is.

> 
> With above approach, both will be 10% of max_caps_per_client. So with current defaults 10k caps by readdir and 1.1M total caps acquired by the client will trigger throttling.

`s/10k/100k/`?

> I think with readdir on large directory kind of workload, decay counter condition will mostly be true and number of total caps ends up deciding readdir throttling.

Yes, it will be mostly true. The client will need to keep up with recall to not get throttled. So ~1.5x mds_max_caps_per_client probably gives it more breathing room.

> In a hypothetical scenario where readdir share of caps is 10k and caps acquired other than readdir 1M, we end up throttling readdir instead of throttling others? Is this ok ?

Reply:
> I don't think there's a clear answer for this without testing. I think keeping the separate config gives us the flexibility to fix this in a production cluster so it makes sense to keep it as-is.

I agree. I will keep it as is for now. 

>  
> `s/10k/100k/`?

Sorry for the typo. Yes it's 100k.